src/HDF5.jl

```diff
@@ -1465,53 +1460,61 @@ Base.axes(dset::Dataset) = map(Base.OneTo, size(dset))
 
 # Write to a subset of a dataset using array slices: dataset[:,:,10] = array
 
-Base.setindex!(dset::Dataset, x, I::Union{AbstractRange,Integer,Colon}...) =
-    _setindex!(dset, x, Base.to_indices(dset, I)...)
-function _setindex!(dset::Dataset, X::Array, I::Union{AbstractRange{Int},Int}...)
-    T = hdf5_to_julia(dset)
-    _setindex!(dset, T, X, I...)
-end
-function _setindex!(dset::Dataset, T::Type, X::Array, I::Union{AbstractRange{Int},Int}...)
-    if !(T <: Array)
-        error("Dataset indexing (hyperslab) is available only for arrays")
-    end
-    ET = eltype(T)
-    if !(ET <: Union{ScalarType, Complex{<:ScalarType}})
-        error("Dataset indexing (hyperslab) is available only for bits types")
-    end
-    if length(X) != prod(map(length, I))
-        error("number of elements in range and length of array must be equal")
+const IndexType = Union{AbstractRange{Int},Int,Colon}
+function Base.setindex!(dset::Dataset, X::Array{T}, I::IndexType...) where T
+    !isconcretetype(T) && error("type $T is not concrete")
+    U = get_jl_type(dset)
+
+    # perform conversions for numeric types
+    if (U <: Number) && (T <: Number) && U !== T
```

Comment:
`Base` doesn't have this restriction, so shouldn't we do the same — just convert if the types don't match?

For example:
```julia
julia> using Test

julia> a = GenericString["a1", "a2"]
2-element Vector{GenericString}:
 "a1"
 "a2"

julia> b = Vector{String}(undef, 2)
2-element Vector{String}:
 #undef
 #undef

julia> b[:] = a
       b
2-element Vector{String}:
 "a1"
 "a2"
```

Reply:
My reasoning here was that `T` and `U` might be technically different types but be structurally the same so that it is unnecessary to require a convert method. For example in the case of compound types `U` will always be a named tuple but `T` could be a user defined struct with the same layout. It seems desirable to support this case without overhead. 

Also, the string example from above already doesn't work with this or in the previous implementation because writing strings requires special handling in order to convert them to a format usable by hdf5. Unifying this in a similar way to #652 but on the write side would be nice but I'm not sure exactly the best way to do it yet.